### ISSUE TITLE
weight_norm ignores magnitude parameter

### ISSUE THREAD

**Issue by nisheethlahoti** [pytorch/pytorch#94759]
### 🐛 Describe the bug

Weight norm completely ignores the contents of `weight_g`. To reproduce, run this snippet:
```python
from torch import nn
from torch.nn.utils.weight_norm import weight_norm

x = nn.Conv2d(2, 2, 3)
weight_norm(x)
x.weight_g.data.zero_()
print(x.weight_g)
print(x.weight_v)
print(x.weight)
```

Actual output: `x.weight_v` and `x.weight` have the same value.
Expected output: `x.weight` should consist entirely of zeros, since `x.weight_g` contains only zeros

### Versions

PyTorch version: 2.0.0.dev20230213
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.0.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: version 3.25.2
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:01:00)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-13.0.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] torch==2.0.0.dev20230213
[conda] pytorch                   2.0.0.dev20230213         py3.9_0    pytorch-nightly

**Comment by stante** [comment]
According to the documentation the weight tensor is calculated before every `forward()` call:

> Weight normalization is implemented via a hook that recomputes the weight tensor from the magnitude and direction before every forward() call.

If you check the weight tensor in your example after you forward through the Conv2d layer, the weight tensor should be all zero.

**Comment by nisheethlahoti** [comment]
Even the behaviour there is off. If I add `x(torch.randn(1, 2, 3, 3))` before the print statements, then the weight tensor is indeed zero. But if I add `x.forward(torch.randn(1, 2, 3, 3))` instead, it's non-zero again.

**Comment by stante** [comment]
`weight_norm` uses the pytorch's hook mechanism and it registers a forward pre hook. In order that these hooks get invoked,  you have to call the module. During the call of a  module any forward pre hooks get invoked, then it calls the module's `forward()` and then invokes any registered forward hooks.

**Comment by albanD** [comment]
Indeed, as mentioned above this implementation relies on the Module's hook to run to do its thing. So this is expected behavior.
The proposal at https://github.com/pytorch/pytorch/issues/7313 will remove this limitation once implemented.
